### PR TITLE
Vaidate URLs

### PR DESCRIPTION
**Pull request by WillAbides**
Resolves google/go-github#1203

This addresses the issue raised in google/go-github#1203 by adding a `newURLString` function that builds a url the same way they were previously built with `fmt.Sprintf` but also errors if the result is not valid.

I went this route instead of modifying `addOptions` because it seemed better to validate all urls instead of only the ones that use `addOptions`.

This turned into a pretty big diff, but that is the nature of updating boilerplate that is used in hundreds of methods.

### CODE REVIEW

File: github/activity_notifications.go [hunk 1]
```diff
@@ -51,8 +50,11 @@ type NotificationListOptions struct {
 //
 // GitHub API docs: https://docs.github.com/en/rest/activity/notifications#list-notifications-for-the-authenticated-user
 func (s *ActivityService) ListNotifications(ctx context.Context, opts *NotificationListOptions) ([]*Notification, *Response, error) {
-	u := "notifications"
-	u, err := addOptions(u, opts)
+	u, err := newURLString("notifications")
+	if err != nil {
+		return nil, nil, err
+	}
```
Comment (gmlewis) [inline github/activity_notifications.go:56]: I don't see any benefit in this case. I think we can safely revert this change.

File: github/activity_star.go [hunk 2]
```diff
@@ -71,12 +73,16 @@ type ActivityListStarredOptions struct {
 // GitHub API docs: https://docs.github.com/en/rest/activity/starring#list-repositories-starred-by-a-user
 func (s *ActivityService) ListStarred(ctx context.Context, user string, opts *ActivityListStarredOptions) ([]*StarredRepository, *Response, error) {
 	var u string
+	var err error
 	if user != "" {
-		u = fmt.Sprintf("users/%v/starred", user)
+		u, err = newURLString("users/%v/starred", user)
 	} else {
-		u = "user/starred"
+		u, err = newURLString("user/starred")
```
Comment (gmlewis) [inline github/activity_star.go:80]: This line can stay as it was before. Any time `newURLString` has only a single constant arg, that change can be reverted.

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @WillAbides .
I'm going to stop review now because I'm curious if you are able to see all the Codecov warnings that I'm able to see?

If not, I can highlight them for you, as there are a lot, as also evidenced by the significant drop in code coverage stats.  It also found some errors, so I'll let you clean those up before I continue reviewing.

Thanks!

**Comment by WillAbides** [comment]
@gmlewis Thanks for having a look. I knew this would likely have a negative impact on code coverage, but decided to wait on the PR for codecov to show me the diff before worrying about it. As an aside now that I've looked at the coverage, 98.1% coverage on master is impressive.

For the urls that are just static strings, I decided to replace those as well so that we are using a consistent pattern across all methods. I don't feel strongly about that and am happy to revert those.

I'm having doubts about the general usefulness of this PR. It seems like if we catch this error in either `newURLString` or `addOptions` we are returning the same error that would have been returned from either [client.NewRequest()](https://github.com/google/go-github/blob/38ca69f37f8597421ceb02e08f2dd225cfaeb1a9/github/github.go#L435-L438) or [http.NewRequest()](https://github.com/golang/go/blob/e68c027204d410ebca5bf1a9660605f2bd737748/src/net/http/request.go#L879-L882). Maybe that wasn't the case when #1203 was filed or maybe there is an exception to this that I'm not seeing.

**Comment by gmlewis** [comment]
> I'm having doubts about the general usefulness of this PR. It seems like if we catch this error in either newURLString or addOptions we are returning the same error that would have been returned from either [client.NewRequest()](https://github.com/google/go-github/blob/38ca69f37f8597421ceb02e08f2dd225cfaeb1a9/github/github.go#L435-L438) or [http.NewRequest()](https://github.com/golang/go/blob/e68c027204d410ebca5bf1a9660605f2bd737748/src/net/http/request.go#L879-L882). Maybe that wasn't the case when https://github.com/google/go-github/issues/1203 was filed or maybe there is an exception to this that I'm not seeing.

Honestly, I am totally fine if you want to call this an interesting experiment and determine that the results are not worth it.

I remember filing #1203 after discovering that an error wasn't triggered when I messed something up which sent alarms through my mind, and was hoping that someone else would give it a good investigation and come to a conclusion, which you graciously have... and thank you!

If you wish to point to this PR in that issue and close them both as "unnecessary", I'm totally fine with that and once again really appreciate your detailed investigation!

**Comment by WillAbides** [comment]
Closing this as unnecessary. I also commented on #1203, but I can't actually close it myself.
